### DISCUSSION TITLE
Let NX explain why it's running what it is running

### DISCUSSION THREAD

**Discussion by alper** [nrwl/nx#26955] [Feature Requests]
### Current Behavior

I have no idea what NX is using as the base for running the stuff it does.

It's entirely opaque and it seems to be hallucinating things.

### Expected Behavior

I would like `--verbose` or something like `nx why` to give me full feedback around why it is running the things it is running.

### Steps to Reproduce

I don't even know because what the hell is it even picking up from where?

### Nx Report

```shell
❯ pnpm nx report                                 [🕙 11 14:06] 

 >  NX   Report complete - copy this into the issue template

   Node   : 18.6.0
   OS     : darwin-arm64
   pnpm   : 8.8.0
   
   nx             : 16.10.0
   @nx/js         : 16.10.0
   @nx/jest       : 16.10.0
   @nx/linter     : 16.10.0
   @nx/workspace  : 16.10.0
   @nx/devkit     : 16.10.0
   @nx/plugin     : 16.10.0
   @nrwl/tao      : 16.10.0
   nx-cloud       : 16.5.2
   typescript     : 4.7.4
   ---------------------------------------
   Community plugins:
   @chocoapp/workspace-plugin : 3.0.29
   ---------------------------------------
   Local workspace plugins:
   	workspace-migration-tool
   	@chocoapp/e2e-plugin
```

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by FrozenPandaz** [comment]
I've converted this into a feature proposal discussion while we work on flushing it out.

I understand you are having difficulty really understanding why Nx runs what it runs. We certainly want to make it easier to understand why Nx does the things it does.

Nx has a feature to append `--graph` to whichever command which runs tasks to show a visual. Running `nx build proj --graph` will show a visual graph of *what* Nx will run. Currently, it is not *why* Nx runs it, but we could add more information on that visual. Seeing the graph will give a good idea of *what* and *how* Nx will run as it is very much what Nx goes by as well.

Could you provide some more concrete cases of when you thought Nx was hallucinating things? Maybe you could highlight the visual and add questions of which parts you don't understand? What kind of information would you expect Nx to show?

**Comment by giltayar** [comment]
> Nx has a feature to append --graph to whichever command which runs tasks to show a visual. Running nx build proj --graph will show a visual graph of what Nx will run.

Not totally correct. If I run a task, and Nx tells me everything comes from the cache and it's actually doing nothing, and now I run it with `--graph`, it will show me a graph of all the dependent tasks, even though it's not running them.

So if I have this use case - I run a task, knowing it _shouldn't_ do anything except this or that dependent task, and it runs other dependent tasks that _shouldn't_ be running, I have no idea what input file caused Nx to unexepectedly run those tasks, and Nx doesn't give me that information.

What I'm (and I believe also @alper) looking for is something like `--dry-run`, where it will tell me - I _will_ be running these tasks. I am running this task because this file changed, and am running this task because a dependent task ran....
